packages/pds/src/api/com/atproto/server/util.ts

```diff
@@ -40,3 +42,71 @@ export const didDocForSession = async (
     dbLogger.warn({ err, did }, 'failed to resolve did doc')
   }
 }
+
+export const isValidDidDocForService = async (
+  ctx: AppContext,
+  did: string,
+): Promise<boolean> => {
+  try {
+    await assertValidDidDocumentForService(ctx, did)
+    return true
+  } catch {
+    return false
+  }
+}
+
+export const assertValidDidDocumentForService = async (
+  ctx: AppContext,
+  did: string,
+) => {
+  if (did.startsWith('did:plc')) {
+    const resolved = await ctx.plcClient.getDocumentData(did)
+    await assertValidDocContents(ctx, did, {
+      pdsEndpoint: resolved.services['atproto_pds']?.endpoint,
+      signingKey: resolved.verificationMethods['atproto'],
+      rotationKeys: resolved.rotationKeys,
+    })
+  } else {
+    const resolved = await ctx.idResolver.did.resolve(did, true)
+    if (!resolved) {
+      throw new InvalidRequestError('Could not resolve DID')
+    }
+    await assertValidDocContents(ctx, did, {
+      pdsEndpoint: getPdsEndpoint(resolved),
+      signingKey: getSigningDidKey(resolved),
+    })
+  }
+}
+
+const assertValidDocContents = async (
+  ctx: AppContext,
+  did: string,
+  contents: {
+    signingKey?: string
+    pdsEndpoint?: string
+    rotationKeys?: string[]
+  },
+) => {
+  const { signingKey, pdsEndpoint, rotationKeys } = contents
+
+  const plcRotationKey =
+    ctx.cfg.entryway?.plcRotationKey ?? ctx.plcRotationKey.did()
+  if (rotationKeys !== undefined && !rotationKeys.includes(plcRotationKey)) {
+    throw new InvalidRequestError(
+      'Server rotation key not included in PLC DID data',
+    )
+  }
+
+  if (!pdsEndpoint || pdsEndpoint !== ctx.cfg.service.publicUrl) {
```

Comment:
If trailing slashes aren't significant, we might consider permitting the pds endpoint both with and without the tailing slash.

Reply:
(If we pursue this, similar logic also appears in `identity.submitPlcOp`.)

Reply:
I'm inclined to just leave it more conservative for now, and we can make it more lax over time if we decide to. Easier to go that way than the other 🤔